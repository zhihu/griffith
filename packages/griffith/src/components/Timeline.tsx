import React, {Component} from 'react'
import {css} from 'aphrodite/no-important'
import Slider, {SliderProps} from './Slider'
import styles, {
  slider as sliderStyles,
  hoveredSlider as hoveredSliderStyles,
  dotHoveredSlider as dotHoveredSliderStyles,
} from './Timeline.styles'

export type TimelineProps = {
  onDragStart?: (...args: any[]) => any
  onDragEnd?: (...args: any[]) => any
  onChange?: (...args: any[]) => any
  onSeek?: (...args: any[]) => any
  onProgressDotHover?: (...args: any[]) => any
  onProgressDotLeave?: (...args: any[]) => any
} & Omit<SliderProps, 'styles'>

export type TimelineState = {
  isHovered: boolean
  isFocused: boolean
  isDragging: boolean
  progressDotHovered: boolean
}

class Timeline extends Component<TimelineProps, TimelineState> {
  state = {
    isHovered: false,
    isFocused: false,
    isDragging: false,
    progressDotHovered: false,
  }

  // refs
  root = null

  handlePointerEnter = () => {
    this.setState({isHovered: true})
  }

  handlePointerLeave = () => {
    this.setState({isHovered: false, isFocused: false})
  }

  handleFocus = () => {
    this.setState({isFocused: true})
  }

  handleBlur = () => {
    this.setState({isFocused: false})
  }

  handleDragStart = () => {
    this.setState({isDragging: true})

    const {onDragStart} = this.props
    if (onDragStart) {
      onDragStart()
    }
  }

  handleDragEnd = () => {
    this.setState({isDragging: false})

    const {onDragEnd} = this.props
    if (onDragEnd) {
      onDragEnd()
    }
  }

  handleChange = (value: any) => {
    const {onSeek, onChange} = this.props
    if (onSeek && value !== (this.props as any).value) {
      onChange?.(value)
    }
    if (this.state.isDragging) return
    if (onSeek) {
      onSeek(value)
    }
  }

  handleProgressDotHover = (...args: any[]) => {
    this.setState({progressDotHovered: true})
    this.props.onProgressDotHover?.(...args)
  }

  handleProgressDotLeave = (...args: any[]) => {
    this.setState({progressDotHovered: false})
    this.props.onProgressDotLeave?.(...args)
  }

  render() {
    const {isHovered, isFocused, isDragging, progressDotHovered} = this.state
    return (
      <div
        className={css(styles.root)}
        onMouseEnter={this.handlePointerEnter}
        onMouseLeave={this.handlePointerLeave}
      >
        <Slider
          {...this.props}
          styles={
            [
              sliderStyles,
              (isHovered || isFocused || isDragging) && hoveredSliderStyles,
              progressDotHovered && dotHoveredSliderStyles,
            ].filter(Boolean) as Record<string, unknown>[]
          }
          onFocus={this.handleFocus}
          onBlur={this.handleBlur}
          onDragStart={this.handleDragStart}
          onDragEnd={this.handleDragEnd}
          onChange={this.handleChange}
          onProgressDotHover={this.handleProgressDotHover}
          onProgressDotLeave={this.handleProgressDotLeave}
        />
      </div>
    )
  }
}

export default Timeline
