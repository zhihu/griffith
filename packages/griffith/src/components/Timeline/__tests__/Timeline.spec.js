import React from 'react'
import {shallow} from 'enzyme'
import Timeline from '..'

describe('Timeline', () => {
  it('get Timeline component', () => {
    const onSeek = jest.fn()
    const onChange = jest.fn()
    const onDragEnd = jest.fn()
    const onDragStart = jest.fn()
    const onProgressDotHover = jest.fn()
    const onProgressDotLeave = jest.fn()
    const wrapper = shallow(
      <Timeline
        buffered={7}
        total={182.234}
        value={0.453}
        onChange={onChange}
        onDragEnd={onDragEnd}
        onDragStart={onDragStart}
        onSeek={onSeek}
        onProgressDotHover={onProgressDotHover}
        onProgressDotLeave={onProgressDotLeave}
      />
    )

    expect(wrapper).toMatchSnapshot()

    expect(wrapper.state().progressDotHovered).toBeFalsy()

    expect(wrapper.state().isHovered).toBeFalsy()

    // PointerEnter
    wrapper.simulate('mouseenter')
    expect(wrapper.state().isHovered).toBeTruthy()

    // PointerLeave
    wrapper.simulate('mouseleave')
    expect(wrapper.state().isHovered).toBeFalsy()

    expect(wrapper.state().isFocused).toBeFalsy()
    // focus event
    wrapper.find('Slider').simulate('focus')
    expect(wrapper.state().isFocused).toBeTruthy()

    // blur event
    wrapper.find('Slider').simulate('blur')
    expect(wrapper.state().isFocused).toBeFalsy()

    expect(wrapper.state().isDragging).toBeFalsy()
    // focus event
    wrapper.find('Slider').simulate('dragstart')
    expect(wrapper.state().isDragging).toBeTruthy()
    expect(onDragStart).toBeCalled()

    // blur event
    wrapper.find('Slider').simulate('dragend')
    expect(wrapper.state().isDragging).toBeFalsy()
    expect(onDragEnd).toBeCalled()

    // handle change
    wrapper.instance().handleChange(47)
    expect(onChange).toBeCalled()
    expect(onSeek).toBeCalled()
  })
})
