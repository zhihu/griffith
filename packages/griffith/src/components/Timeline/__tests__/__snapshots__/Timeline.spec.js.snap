// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline get Timeline component 1`] = `
<div
  className=""
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <Slider
    buffered={7}
    onBlur={[Function]}
    onChange={[Function]}
    onDragEnd={[Function]}
    onDragStart={[Function]}
    onFocus={[Function]}
    onSeek={[MockFunction]}
    orientation="horizontal"
    progressDots={Array []}
    reverse={false}
    step={1}
    styles={
      Array [
        Object {
          "bar": Object {
            "_definition": Object {
              "backgroundColor": "#0f88eb",
            },
            "_len": 29,
            "_name": "bar_1kvrquw",
          },
          "buffered": Object {
            "_definition": Object {
              "backgroundColor": "rgba(255, 255, 255, 0.6)",
              "borderRadius": "2px 0 0 2px",
            },
            "_len": 75,
            "_name": "buffered_gsgvhl",
          },
          "root": Object {
            "_definition": Object {
              "height": 40,
            },
            "_len": 13,
            "_name": "root_1pr59fm",
          },
          "thumb": Object {
            "_definition": Object {
              "transform": "scale(0)",
              "transition": "transform 300ms",
            },
            "_len": 55,
            "_name": "thumb_1omjhz8",
          },
          "track": Object {
            "_definition": Object {
              "transition": "transform 300ms",
            },
            "_len": 32,
            "_name": "track_1q3h7g7",
          },
        },
        false,
      ]
    }
    total={182.234}
    value={0.453}
  />
</div>
`;
